### PR TITLE
MAINT: Provide prefetch history loader for 1m.

### PR DESCRIPTION
**Pull request by ehebert**
Factor out USEquityHistoryLoader into a base class which can be
overridden to support daily or minute history.

In support of reading the minute bars for a range of market minute by
the loader from the reader, put the entire minute index for the minute
bar reader into memory so that the two indexes can be readily compared.
(Though we may want to consider a better way to store just the half day
 information, since the half day only accounts for ~1% of trading days.)

Change some test case sim params so that the tests are started far
enough from the dataset start so that history calls can be made.

### CODE REVIEW

File: zipline/data/_minute_bar_internal.pyx [hunk 1]
```diff
@@ -59,15 +59,10 @@ def find_position_of_minute(ndarray[long_t, ndim=1] market_opens,
     """
     cdef Py_ssize_t market_open_loc, market_open, delta
 
-    market_open_loc = \
-        searchsorted(market_opens, minute_val, side='right') - 1
-    market_open = market_opens[market_open_loc]
-    delta = minute_val - market_open
-
-    return (market_open_loc * minutes_per_day) + delta
+    return searchsorted(minutes, minute_val, side='right') - 1
```
Comment (jbredeche) [inline zipline/data/_minute_bar_internal.pyx]: you don't need the variables defined in line 60 anymore

File: zipline/data/_minute_bar_internal.pyx [hunk 2]
```diff
@@ -100,17 +95,15 @@ def find_last_traded_position_internal(
     """
     cdef Py_ssize_t minute_pos, current_minute
 
-    minute_pos = int_min(
-        find_position_of_minute(market_opens, end_minute, minutes_per_day),
+    start_pos = searchsorted(minutes, start_minute, side='left')
```
Comment (jbredeche) [inline zipline/data/_minute_bar_internal.pyx]: you should `cdef` `start_pos` and `end_pos` above, right? (and get rid of `current_minute`)

File: zipline/data/_minute_bar_internal.pyx [hunk 3]
```diff
@@ -100,17 +95,15 @@ def find_last_traded_position_internal(
...
+
+    end_pos = minute_pos = int_min(
```
Comment (jbredeche) [inline zipline/data/_minute_bar_internal.pyx]: does `end_pos` still need to exist?

File: zipline/data/data_portal.py [hunk 4]
```diff
@@ -495,6 +501,15 @@ def __init__(self,
             'price': None
         }
         self._equity_daily_reader_array_data = {}
+        self._equity_minute_loader_array_keys = {
```
Comment (jbredeche) [inline zipline/data/data_portal.py]: can you add a comment explaining what's going to be stored in `self._equity_minute_loader_array_keys` and `self._equity_minute_loader_array_data`?  I assume it's cached data of some sort, but hard to tell.

File: zipline/data/data_portal.py [hunk 5]
```diff
@@ -1441,17 +1365,11 @@ def _apply_all_adjustments(self, data, asset, dts, field,
...
 
-            data *= price_adj_factor
-
-            # if anything is zero, it's a missing bar, so replace it with NaN.
-            # we only want to do this for non-volume fields, because a missing
-            # volume should be 0.
-            data[data == 0] = np.NaN
-
-        np.around(data, 3, out=data)
+            if price_adj_factor is not None:
```
Comment (jbredeche) [inline zipline/data/data_portal.py]: this is likely going to conflict with lazy-mainline, where this method has been renamed and tweaked a bit.

File: zipline/data/minute_bars.py [hunk 6]
```diff
@@ -612,31 +610,26 @@ def get_last_traded_dt(self, asset, dt):
...
-            US_EQUITIES_MINUTES_PER_DAY
-        )
-
-        return pd.Timestamp(minute_epoch, tz='UTC', unit="m")
+        minute_epoch = self._minute_value_index[pos]
+        return pd.Timestamp(minute_epoch, tz='UTC')
 
     @remember_last
-    def _find_position_of_minute(self, minute_dt):
+    def _find_position_of_minute(self, minute_dt_val):
```
Comment (jbredeche) [inline zipline/data/minute_bars.py]: can you update the docstring,`minute_dt_val` is the nanosecond value right?

File: zipline/data/minute_bars.py [hunk 7]
```diff
@@ -706,3 +698,16 @@ def unadjusted_window(self, fields, start_dt, end_dt, sids):
                 out *= self._ohlc_inverse
             results.append(out)
         return results
+
+    def unadjusted_window_for_minutes(self, fields, minutes, sids):
```
Comment (jbredeche) [inline zipline/data/minute_bars.py]: can you add a docstring here?  hard to decipher what this method does.

File: zipline/data/minute_bars.py [hunk 8]
```diff
@@ -706,3 +698,16 @@ def unadjusted_window(self, fields, start_dt, end_dt, sids):
...
+        start_dt = minutes[0]
+        end_dt = minutes[-1]
+        results = self.unadjusted_window(fields, start_dt, end_dt, sids)
+        if len(results[0][0]) != len(minutes):
+            start_idx = self._find_position_of_minute(start_dt.value)
+            end_idx = self._find_position_of_minute(end_dt.value)
+            data_minutes = self._minute_value_index[start_idx:end_idx + 1]
+            indices = np.in1d(data_minutes, minutes.asi8)
+            # Should do this to ndarray.
+            results = [r[:, indices] for r in results]
```
Comment (jbredeche) [inline zipline/data/minute_bars.py]: just to make sure I understand, this entire block handles early closes?  can you add some comments to that effect?

File: zipline/data/minute_bars.py [hunk 9]
```diff
@@ -706,3 +698,16 @@ def unadjusted_window(self, fields, start_dt, end_dt, sids):
...
+
+    def unadjusted_window_for_minutes(self, fields, minutes, sids):
+        start_dt = minutes[0]
+        end_dt = minutes[-1]
+        results = self.unadjusted_window(fields, start_dt, end_dt, sids)
+        if len(results[0][0]) != len(minutes):
+            start_idx = self._find_position_of_minute(start_dt.value)
+            end_idx = self._find_position_of_minute(end_dt.value)
+            data_minutes = self._minute_value_index[start_idx:end_idx + 1]
+            indices = np.in1d(data_minutes, minutes.asi8)
```
Comment (jbredeche) [inline zipline/data/minute_bars.py]: from the docs, it seems like `np.in1d` is O(len(data_minutes) * len(minutes)) - is that fast enough for our use here, or am I misreading?

Comment (jbredeche) [inline zipline/data/minute_bars.py]: nvm, https://github.com/numpy/numpy/blob/v1.10.1/numpy/lib/arraysetops.py#L305-L404 shows that it's sorting the arrays first, phew.

Comment (jbredeche) [inline zipline/data/minute_bars.py]: actually, there's an `assume_unique` flag that I think we can set to True, which would make this faster: https://github.com/numpy/numpy/blob/v1.10.1/numpy/lib/arraysetops.py#L318

File: zipline/data/us_equity_loader.py [hunk 10]
```diff
@@ -121,7 +136,7 @@ def _get_adjustments_in_range(self, assets, days, field):
                 for m in mergers:
                     dt = m[0]
                     if start < dt <= end:
-                        end_loc = days.get_loc(dt)
+                        end_loc = dts.searchsorted(dt)
```
Comment (jbredeche) [inline zipline/data/us_equity_loader.py]: why this change?  because `dts` can be minutes now?

File: zipline/data/us_equity_loader.py [hunk 11]
```diff
@@ -179,50 +195,25 @@ def _ensure_sliding_window(
...
-                end_ix = self._calendar.get_loc(end)
-            if field != 'volume':
-                pre_array = full((fill_size, 1), nan)
-            else:
-                pre_array = full((fill_size, 1), 0)
-        else:
-            offset = 0
-            start_ix = self._calendar.get_loc(start)
-            end_ix = self._calendar.get_loc(end)
+        start = dts[0]
```
Comment (jbredeche) [inline zipline/data/us_equity_loader.py]: style comment/question, why are `start` and `end` defined far away from each other?  easier to read if they're defined right next to each other, I think.

File: zipline/data/us_equity_loader.py (hunk 10 above)
Comment (jbredeche) [inline zipline/data/us_equity_loader.py]: out of scope for this PR, but would love to see less duplicates of the intricate logic around each merger, dividend, and split (all the `Float64Multiply` construction, etc).  I'll submit a PR afterwards.

File: zipline/data/us_equity_loader.py [hunk 12]
```diff
@@ -168,10 +183,11 @@ def _get_adjustments_in_range(self, assets, days, field):
         return adjs
 
     def _ensure_sliding_window(
-            self, assets, start, end, size, field):
+            self, assets, dts, size, field):
```
Comment (jbredeche) [inline zipline/data/us_equity_loader.py]: I have no idea what this method does.  can you add some comments explaining?

File: zipline/data/us_equity_loader.py [hunk 13]
```diff
@@ -179,50 +195,25 @@ def _ensure_sliding_window(
...
-                pre_array = full((fill_size, 1), nan)
-            else:
-                pre_array = full((fill_size, 1), 0)
-        else:
-            offset = 0
-            start_ix = self._calendar.get_loc(start)
-            end_ix = self._calendar.get_loc(end)
+        start = dts[0]
+
+        offset = 0
```
Comment (jbredeche) [inline zipline/data/us_equity_loader.py]: `offset` is only ever assigned 0, then passed into `SlidingWindow` below - can we get rid of it?

File: tests/test_history.py [hunk 14]
```diff
@@ -1255,7 +1260,7 @@ def test_daily_splits_and_mergers(self):
                 "close"
             )[asset]
 
-            # first value should be halved, second value unadjusted
+            # first value should be doubled, second value unadjusted
```
Comment (jbredeche) [inline tests/test_history.py]: still looks like it's getting halved (2 -> 1), is this comment change intentional?

File: zipline/data/data_portal.py (hunk 5 above)
Comment (ehebert) [inline zipline/data/data_portal.py]: I do not see that rename on current `lazy-mainline`.

File: zipline/data/minute_bars.py (hunk 9 above)
Comment (ehebert) [inline zipline/data/minute_bars.py]: Tuning this would be good, howevere it is also called relatively infrequently. Once per every 1490 minute bars.

File: zipline/data/minute_bars.py (hunk 8 above)
Comment (ehebert) [inline zipline/data/minute_bars.py]: Yes, I am having some pause on how the half days are handled here.
I believe we can do it instead by just storing the minutes that *aren't* market minutes and filtering those out; since half days are only 2-3 times a year we should avoid some of the upfront costs that loading all minutes adds.

In that case, would revert the changes that change `_minute_bar_internal.pyx`.

File: tests/test_history.py (hunk 14 above)
Comment (ehebert) [inline tests/test_history.py]: You're right, that comment was meant to go with lines 1276 and 1279.

File: zipline/data/us_equity_loader.py (hunk 10 above)
Comment (ehebert) [inline zipline/data/us_equity_loader.py]: Yes, because they can be `dts`.

I am planning on factoring this out as well.

File: zipline/data/us_equity_loader.py (hunk 11 above)
Comment (ehebert) [inline zipline/data/us_equity_loader.py]: The reason that it is not, is if the previous window has not expired, `start` is not needed.

File: zipline/data/us_equity_loader.py (hunk 13 above)
Comment (ehebert) [inline zipline/data/us_equity_loader.py]: Going the other direction, and passing `offset` to `Float64Window` as well.

**Comment by ehebert** [comment]
Closing in favor of https://github.com/quantopian/zipline/pull/1091
